README.md
# cron-lingo

A small Rust library to parse a cron-like, human-readable expression like "at 6 AM on Mondays and at 6 PM (Saturdays and Sundays)" and use it to iterate upcoming dates.

The main goal is to provide a more predictable way for e.g. schedulling critical tasks by getting rid of some core functionality of standard cron. Also the expression syntax is self-explanatory to a large extent, which may present a useful side-effect if you are planning to expose the configuration of some scheduler to non-technical staff.

## Example

```rust
use cron_lingo::Schedule;
use std::str::FromStr;

fn main() {
    let schedule = Schedule::from_str("at 1 PM on Mondays").unwrap();

    for date in schedule.iter().unwrap().take(3) {
        println!("{}", date);
    }
}

// Output:
// 2021-06-14 13:00 +2
// 2021-06-21 13:00 +2
// 2021-06-28 13:00 +2
```

Please check out the module-level documentation on [docs.rs](https://docs.rs/cron-lingo) for specifics on the applied syntax.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.4.2] - 2022-03-12
### Added
- impl From<Schedule> for MultiSchedule

## [0.4.1] - 2022-03-08
### Added
- Added methods `ScheduleIter::assume_offset`, `ScheduleIter::use_local_offset` and pendants for `MultiScheduleIter` that allow to change the default offset (local) that dates are computed in to another.

## [0.4.0] - 2021-12-04
### Added
- Introduced new WeekdayModifier variant "Last" to select the last occurrence of some weekday, e.g. "at 6 PM (last Monday)".
- Added `MultiSchedule` to accomodate a combination of several distinct `Schedule`s.
- Added `cron_lingo::error::Error::UnexpectedEndOfInput`.
- Added `cron_lingo::error::Error::Syntax` and `cron_lingo::error::SyntaxError`.
### Changed
- Renamed `cron_lingo::error::InvalidExpression` to simply `cron_lingo::error::Error`.
- Expressions must now be parsed separately using `from_str` but may then be combined together with `std::ops::Add` or added to an existing `MultiSchedule` using `std::ops::AddAssign`.
- Iterator of `Schedule` or `MultiSchedule` is now fallible (`next()` returns `Option<Result<time::OffsetDateTime, cron_lingo::error::Error>`) as the retrieval of the local offset may fail.
- Creation of said iterators via `Schedule::iter()` or `MultiSchedule::iter()` is now fallible as well for the same reason.
- Moved from time v0.2.22 to v0.3. Compile flag is needed in order to avoid runtime errors when the local offset is retrieved internally: `RUSTFLAGS="--cfg unsound_local_offset" cargo build`. The compile flag may be omitted when your application is single-threaded.
### Removed
- Removed variants `DuplicateInput` and `IllogicalWeekdayCombination` from error enum as the occurrence of duplicates et al. does not prevent correct computation of the next upcoming date.

## [0.3.1] - 2021-06-10
### Added
- Check for illogical combinations of weekdays and modifiers, e.g. "on Tuesdays and the 3rd Tuesday"
### Changed
- Updated module description for more clarity regarding expression syntax

## [0.3.0] - 2021-06-09
### Added
- Add `skip_outdated()` method to ScheduleIter to override default behaviour in `next()`. By default `next()` will never return a date that is in the past, but instead resume the iteration from the current local time.
### Changed
- Rename Timetable to Schedule
- Complete expression syntax overhaul to offer more possibilities

## [0.2.2] - 2021-05-25
### Changed
- Remove explicit lifetime from TimetableIter in order to use it with pyo3

## [0.2.1] - 2021-03-16
### Changed
- Fixed readme to reflect latest changes ...

## [0.2.0] - 2021-03-15
### Changed
- `Timetable` does not implement `Iterator` anymore. Instead `.iter()` returns a wrapper struct `TimetableIter` that keeps track of state during iteration.
- Removed function `new()`.


tests/parse.rs
use cron_lingo::Schedule;
use std::str::FromStr;

#[test]
fn test_empty_expression() {
    let result = Schedule::from_str("").unwrap_err();
    assert_eq!(result, cron_lingo::error::Error::EmptyExpression);
}

#[test]
fn test_unexpected_end_of_input() {
    let result = Schedule::from_str("at").unwrap_err();
    assert_eq!(result, cron_lingo::error::Error::UnexpectedEndOfInput);

    let result = Schedule::from_str("at ").unwrap_err();
    assert_eq!(result, cron_lingo::error::Error::UnexpectedEndOfInput);

    let result = Schedule::from_str("at 08").unwrap_err();
    assert_eq!(result, cron_lingo::error::Error::UnexpectedEndOfInput);

    let result = Schedule::from_str("at 8:").unwrap_err();
    assert_eq!(result, cron_lingo::error::Error::UnexpectedEndOfInput);

    let result = Schedule::from_str("at 8:0").unwrap_err();
    assert_eq!(result, cron_lingo::error::Error::UnexpectedEndOfInput);

    let result = Schedule::from_str("at 8:00 ").unwrap_err();
    assert_eq!(result, cron_lingo::error::Error::UnexpectedEndOfInput);

    let result = Schedule::from_str("at 8:00 AM ").unwrap_err();
    assert_eq!(result, cron_lingo::error::Error::UnexpectedEndOfInput);

    let result = Schedule::from_str("at 8:00 AM on").unwrap_err();
    assert_eq!(result, cron_lingo::error::Error::UnexpectedEndOfInput);

    let result = Schedule::from_str("at 6 AM on Mondays and").unwrap_err();
    assert_eq!(result, cron_lingo::error::Error::UnexpectedEndOfInput);

    let result = Schedule::from_str("at 6 AM on Mondays and Thursdays ").unwrap_err();
    assert_eq!(result, cron_lingo::error::Error::UnexpectedEndOfInput);

    let result = Schedule::from_str("at 6 AM (").unwrap_err();
    assert_eq!(result, cron_lingo::error::Error::UnexpectedEndOfInput);

    let result = Schedule::from_str("at 6 AM (Mondays").unwrap_err();
    assert_eq!(result, cron_lingo::error::Error::UnexpectedEndOfInput);

    let result = Schedule::from_str("at 6 AM (Mondays and").unwrap_err();
    assert_eq!(result, cron_lingo::error::Error::UnexpectedEndOfInput);

    let result = Schedule::from_str("at 6 AM (Mondays and Thursdays").unwrap_err();
    assert_eq!(result, cron_lingo::error::Error::UnexpectedEndOfInput);

    let result = Schedule::from_str("at 6 AM (Mondays and Thursdays) ").unwrap_err();
    assert_eq!(result, cron_lingo::error::Error::UnexpectedEndOfInput);
}

#[test]
fn test_schedule_1() {
    let expr = "at 6 AM (Mondays and Thursdays)";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_2() {
    let expr = "at 6 PM on Saturdays and Sundays";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_3() {
    let expr = "at 6 AM on the last Monday";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_4() {
    let expr = "at 6 AM, 6 PM (Mondays) in even weeks";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_5() {
    let expr = "at 2 PM (Mondays, Thursdays, Saturdays and last Sunday) in even weeks";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_6() {
    let expr = "at 6:30 AM on the first Monday and 4th Saturday";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_7() {
    let expr = "at 8:15 AM (Fridays and the first Saturday) in even weeks";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_8() {
    let expr = "at 6 PM on Sundays in even weeks";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_9() {
    let expr = "at 6 PM on the 1st Saturday and Sundays";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_10() {
    let expr = "at 6 PM on Thursdays in odd weeks";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_11() {
    let expr = "at 8 AM and 8 PM on the first Sunday";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_12() {
    let expr = "at 1 AM, 5:30 AM, 12 PM, 5:30 PM and 11:59 PM";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_13() {
    let expr = "at 6:30 PM on Thursdays";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}


tests/schedule.rs
use cron_lingo::Schedule;
use std::str::FromStr;

#[test]
fn test_empty_expression() {
    let result = Schedule::from_str("").unwrap_err();
    assert_eq!(result, cron_lingo::error::Error::EmptyExpression);
}

#[test]
fn test_unexpected_end_of_input() {
    let expr = "at 6 AM on Mondays and";
    let result = Schedule::from_str(expr).unwrap_err();
    assert_eq!(result, cron_lingo::error::Error::UnexpectedEndOfInput);
}

#[test]
fn test_schedule_1() {
    let expr = "at 6 AM (Mondays and Thursdays)";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_2() {
    let expr = "at 6 PM on Saturdays and Sundays";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_3() {
    let expr = "at 6 AM on the last Monday";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_4() {
    let expr = "at 6 AM, 6 PM (Mondays) in even weeks";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_5() {
    let expr = "at 2 PM (Mondays, Thursdays, Saturdays and last Sunday) in even weeks";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_6() {
    let expr = "at 6:30 AM on the first Monday and 4th Saturday";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_7() {
    let expr = "at 8:15 AM (Fridays and the first Saturday) in even weeks";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_8() {
    let expr = "at 6 PM on Sundays in even weeks";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_9() {
    let expr = "at 6 PM on the 1st Saturday and Sundays";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_10() {
    let expr = "at 6 PM on Thursdays in odd weeks";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_11() {
    let expr = "at 8 AM and 8 PM on the first Sunday";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_12() {
    let expr = "at 1 AM, 5:30 AM, 12 PM, 5:30 PM and 11:59 PM";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}

#[test]
fn test_schedule_13() {
    let expr = "at 6:30 PM on Thursdays";
    let result = Schedule::from_str(expr);
    assert!(result.is_ok(), "{:?}", result);
}


tests/timetable.rs
use cron_lingo::Timetable;
use std::str::FromStr;

#[test]
fn test_timetable() {
    let expr = "at 17 o'clock on Tuesday and Thursday in odd weeks";
    assert!(Timetable::from_str(expr).is_ok());
}


src/error.rs
use std::error::Error as StdError;
use std::fmt;

/// A global error type that encapsulates all other, more specific
/// error types.
#[non_exhaustive]
#[derive(Debug, PartialEq)]
pub enum Error {
    EmptyExpression,
    Syntax(SyntaxError),
    UnexpectedEndOfInput,
    TimeParse(time::error::Parse),
    IndeterminateOffset(time::error::IndeterminateOffset),
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Self::EmptyExpression => write!(f, "the expression string must not be empty"),
            Self::Syntax(e) => e.fmt(f),
            Self::UnexpectedEndOfInput => write!(
                f,
                "the parser reached the end of the expression but expected more characters"
            ),
            Self::TimeParse(e) => write!(f, "failed to parse time: {}", e),
            Self::IndeterminateOffset(e) => e.fmt(f),
        }
    }
}

impl StdError for Error {}

/// Generic syntax error. Gives the exact position of the erroneous character
/// in an expression and points out the expected input.
#[derive(Debug, Clone, PartialEq)]
pub struct SyntaxError {
    pub(crate) position: usize,
    pub(crate) expected: String,
    pub(crate) continues: String,
}

impl fmt::Display for SyntaxError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "unexpected sequence of characters starting at position '{}', expected {}, got '{}'",
            self.position, self.expected, self.continues
        )
    }
}

impl StdError for SyntaxError {}


src/parse.rs
use crate::error::*;
use crate::types::{ParsedSchedule, WeekVariant, WeekdayModifier};
use time::{Time, Weekday};

// Prepares a format description for times formatted as e.g. "1 AM" or "01 AM".
const TIME_FORMAT_NO_MINUTES: &[time::format_description::FormatItem] =
    time::macros::format_description!("[hour padding:none repr:12] [period case:upper]");

// Prepares a format description for times formatted as e.g. "1:00 AM" or "01:30 AM".
const TIME_FORMAT_WITH_MINUTES: &[time::format_description::FormatItem] =
    time::macros::format_description!("[hour padding:none repr:12]:[minute] [period case:upper]");

// Parses an expression block by block which are concatenated by "plus", checking for
// possibly reaching the end of the expression along the way.
// Returns a collection of parsed blocks.
pub(crate) fn parse(expression: &str) -> Result<ParsedSchedule, Error> {
    let chars: Vec<char> = expression.chars().collect();
    let mut position: usize = 0;

    if chars.is_empty() {
        return Err(Error::EmptyExpression);
    }

    eat_keyword("at", &mut position, &chars)?;
    eat_whitespace(&mut position, &chars)?;

    let times = match_times(&mut position, &chars)?;

    let days = if position < chars.len() {
        Some(match_weekdays(&mut position, &chars)?)
    } else {
        None
    };

    let weeks = if position < chars.len() {
        eat_whitespace(&mut position, &chars)?;
        Some(match_week(&mut position, &chars)?)
    } else {
        None
    };

    let spec = ParsedSchedule { times, days, weeks };

    Ok(spec)
}

// Looks ahead and checks for an arbitrary pattern without actually advancing the
// pointer.
fn expect_sequence(sequence: &str, position: &usize, chars: &[char]) -> bool {
    let end_pos = *position + sequence.len();
    match chars.get(*position..end_pos) {
        Some(c) => c.iter().collect::<String>().as_str() == sequence,
        None => false,
    }
}

// Consumes a pattern, usually a single word, by advancing the pointer beyond the
// end of the pattern.
// Returns an error if there are no more characters to match against or the actual
// collection of characters does not match the pattern.
fn eat_keyword(keyword: &str, position: &mut usize, chars: &[char]) -> Result<(), Error> {
    let end_pos = *position + keyword.len();

    let word: String = chars
        .get(*position..end_pos)
        .ok_or(Error::UnexpectedEndOfInput)?
        .iter()
        .collect();

    if word.as_str() == keyword {
        *position = end_pos;
    } else {
        let err = SyntaxError {
            position: *position,
            expected: format!("'{}'", keyword),
            continues: chars
                .get(*position..*position + 10)
                .or(chars.get(*position..))
                .unwrap()
                .iter()
                .collect::<String>(),
        };
        return Err(Error::Syntax(err));
    }

    Ok(())
}

// Consumes a well-defined weekday modifier and returns a parsed representation of it
// or an error if the pattern at hand does not match the expected string.
fn eat_modifier(position: &mut usize, chars: &[char]) -> Result<WeekdayModifier, Error> {
    if eat_keyword("1st", position, chars).is_ok() {
        return Ok(WeekdayModifier::First);
    }

    if eat_keyword("first", position, chars).is_ok() {
        return Ok(WeekdayModifier::First);
    }

    if eat_keyword("2nd", position, chars).is_ok() {
        return Ok(WeekdayModifier::Second);
    }

    if eat_keyword("second", position, chars).is_ok() {
        return Ok(WeekdayModifier::Second);
    }

    if eat_keyword("3rd", position, chars).is_ok() {
        return Ok(WeekdayModifier::Third);
    }

    if eat_keyword("third", position, chars).is_ok() {
        return Ok(WeekdayModifier::Third);
    }

    if eat_keyword("4th", position, chars).is_ok() {
        return Ok(WeekdayModifier::Fourth);
    }

    if eat_keyword("fourth", position, chars).is_ok() {
        return Ok(WeekdayModifier::Fourth);
    }

    if eat_keyword("last", position, chars).is_ok() {
        return Ok(WeekdayModifier::Last);
    }

    let err = SyntaxError {
        position: *position,
        expected:
            "one of '1st', 'first', '2nd', 'second', '3rd', 'third', '4th', 'fourth' or 'last'"
                .to_string(),
        continues: chars
            .get(*position..*position + 10)
            .or(chars.get(*position..))
            .unwrap()
            .iter()
            .collect::<String>(),
    };

    Err(Error::Syntax(err))
}

// Consumes a well-defined weekday (either worded in a specific or "general" way).
// Returns either the parsed representation of and error if the pattern at hand
// does not match the expected string or no more characters are there to consume.
fn eat_weekday(position: &mut usize, chars: &[char], specific: bool) -> Result<Weekday, Error> {
    let day;

    if eat_keyword("Monday", position, chars).is_ok() {
        day = Weekday::Monday;
    } else if eat_keyword("Tuesday", position, chars).is_ok() {
        day = Weekday::Tuesday;
    } else if eat_keyword("Wednesday", position, chars).is_ok() {
        day = Weekday::Wednesday;
    } else if eat_keyword("Thursday", position, chars).is_ok() {
        day = Weekday::Thursday;
    } else if eat_keyword("Friday", position, chars).is_ok() {
        day = Weekday::Friday;
    } else if eat_keyword("Saturday", position, chars).is_ok() {
        day = Weekday::Saturday;
    } else if eat_keyword("Sunday", position, chars).is_ok() {
        day = Weekday::Sunday;
    } else {
        let err = SyntaxError {
            position: *position,
            expected: "one of 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday' or 'Sunday'".to_string(),
            continues: chars
                .get(*position..*position + 10)
                .or(chars.get(*position..))
                .unwrap()
                .iter()
                .collect::<String>(),
        };
        return Err(Error::Syntax(err));
    }

    if !specific {
        if let Some(c) = chars.get(*position) {
            if *c == 's' {
                *position += 1;
                return Ok(day);
            } else {
                let err = SyntaxError {
                        position: *position,
                        expected: "one of 'Mondays', 'Tuesdays', 'Wednesdays', 'Thursdays', 'Fridays', 'Saturdays' or 'Sundays'".to_string(),
                        continues: chars
                            .get(*position..*position + 10)
                            .or(chars.get(*position..))
                            .unwrap()
                            .iter()
                            .collect::<String>(),
                    };
                return Err(Error::Syntax(err));
            }
        } else {
            return Err(Error::UnexpectedEndOfInput);
        }
    }

    return Ok(day);
}

// Consumes a single whitespace or returns an error.
fn eat_whitespace(position: &mut usize, chars: &[char]) -> Result<(), Error> {
    let ch = chars.get(*position).ok_or(Error::UnexpectedEndOfInput)?;

    if ch.is_whitespace() {
        *position += 1;
        Ok(())
    } else {
        let err = SyntaxError {
            position: *position,
            expected: "a whitespace".to_string(),
            continues: chars
                .get(*position..*position + 10)
                .or(chars.get(*position..))
                .unwrap()
                .iter()
                .collect::<String>(),
        };
        Err(Error::Syntax(err))
    }
}

// Matches, parses and returns a collection of parsed times.
fn match_times(position: &mut usize, chars: &[char]) -> Result<Vec<Time>, Error> {
    let mut tokens = vec![];

    tokens.push(match_time(position, chars)?);

    // Check for more occurrences of time tokens.
    loop {
        match chars.get(*position) {
            Some(ch) => {
                if *ch == ',' {
                    *position += 1;
                    eat_whitespace(position, chars)?;
                    tokens.push(match_time(position, chars)?);
                    continue;
                } else if ch.is_whitespace() {
                    if expect_sequence(" and", &position, &chars) {
                        eat_whitespace(position, chars)?;
                        eat_keyword("and", position, chars)?;
                        eat_whitespace(position, chars)?;
                        tokens.push(match_time(position, chars)?);
                        continue;
                    } else {
                        break;
                    }
                } else {
                    let err = SyntaxError {
                        position: *position,
                        expected: "either ',' or a whitespace".to_string(),
                        continues: chars
                            .get(*position..*position + 10)
                            .or(chars.get(*position..))
                            .unwrap()
                            .iter()
                            .collect::<String>(),
                    };
                    return Err(Error::Syntax(err));
                }
            }
            None => break,
        }
    }

    Ok(tokens)
}

// Matches and parses a single time.
fn match_time(position: &mut usize, chars: &[char]) -> Result<Time, Error> {
    // First character must be a number.
    let hour = chars
        .get(*position)
        .ok_or(Error::UnexpectedEndOfInput)?
        .clone();

    if !hour.is_numeric() {
        let err = SyntaxError {
            position: *position,
            expected: "a number in the range 01..=12".to_string(),
            continues: chars
                .get(*position..*position + 10)
                .or(chars.get(*position..))
                .unwrap()
                .iter()
                .collect::<String>(),
        };
        return Err(Error::Syntax(err));
    }

    *position += 1;

    // Next character may be the next part of a 2-digit number, a colon,
    // or a whitespace.
    let next = chars.get(*position).ok_or(Error::UnexpectedEndOfInput)?;

    if next.is_whitespace() {
        *position += 1;
        let mut time = String::new();
        time.push(hour);
        time.push(' ');

        for i in 0..2 {
            time.push(match_period(i, position, chars)?);
        }

        *position += 2;

        let parsed = Time::parse(&time, &TIME_FORMAT_NO_MINUTES).map_err(Error::TimeParse)?;

        Ok(parsed)
    } else if *next == ':' {
        *position += 1;
        let mut time = String::new();
        time.push(hour);
        time.push(*next);

        for i in 0..2 {
            time.push(match_minute(i, position, chars)?);
        }

        *position += 2;

        eat_whitespace(position, chars)?;
        time.push(' ');

        for i in 0..2 {
            time.push(match_period(i, position, chars)?);
        }

        *position += 2;

        let parsed = Time::parse(&time, &TIME_FORMAT_WITH_MINUTES).map_err(Error::TimeParse)?;

        Ok(parsed)
    } else if next.is_numeric() {
        *position += 1;
        let mut time = String::new();
        time.push(hour);
        time.push(*next);

        let next = chars
            .get(*position)
            .ok_or(Error::UnexpectedEndOfInput)?
            .clone();

        *position += 1;

        if next.is_whitespace() {
            time.push(' ');

            for i in 0..2 {
                time.push(match_period(i, position, chars)?);
            }

            *position += 2;

            let parsed = Time::parse(&time, &TIME_FORMAT_NO_MINUTES).map_err(Error::TimeParse)?;

            Ok(parsed)
        } else if next == ':' {
            time.push(next);

            for i in 0..2 {
                time.push(match_minute(i, position, chars)?);
            }

            *position += 2;

            eat_whitespace(position, chars)?;
            time.push(' ');

            for i in 0..2 {
                time.push(match_period(i, position, chars)?);
            }

            *position += 2;

            let parsed = Time::parse(&time, &TIME_FORMAT_WITH_MINUTES).map_err(Error::TimeParse)?;

            Ok(parsed)
        } else {
            let err = SyntaxError {
                position: *position,
                expected: "either ',' or a whitespace".to_string(),
                continues: chars
                    .get(*position..*position + 10)
                    .or(chars.get(*position..))
                    .unwrap()
                    .iter()
                    .collect::<String>(),
            };
            Err(Error::Syntax(err))
        }
    } else {
        let err = SyntaxError {
            position: *position,
            expected: "the remainder of a number in the range 01..=12, a colon or a whitespace"
                .to_string(),
            continues: chars
                .get(*position..*position + 10)
                .or(chars.get(*position..))
                .unwrap()
                .iter()
                .collect::<String>(),
        };
        Err(Error::Syntax(err))
    }
}

// Match and check the next character as part of a double-digit time period
// on a surface level (heavy lifting is done by the time crate).
// In order to provide a more useful error message this function only checks one
// part of a compound using an extra index that is added to the position. When
// the function fails the position points to the start of the compound instead
// of the single offending character.
fn match_period(index: usize, position: &mut usize, chars: &[char]) -> Result<char, Error> {
    let c = chars
        .get(*position + index)
        .ok_or(Error::UnexpectedEndOfInput)?;

    if c.is_alphabetic() && c.is_uppercase() {
        Ok(*c)
    } else {
        let err = SyntaxError {
            position: *position + index,
            expected: "either 'AM or 'PM'".to_string(),
            continues: chars
                .get(*position..*position + 10)
                .or(chars.get(*position..))
                .unwrap()
                .iter()
                .collect::<String>(),
        };
        return Err(Error::Syntax(err));
    }
}

// Match and check the next character as part of a double-digit minute compound
// on a surface level (heavy lifting is done by the time crate).
// In order to provide a more useful error message this function only checks one
// part of a compound using an extra index that is added to the position. When
// the function fails the position points to the start of the compound instead
// of the single offending character.
fn match_minute(index: usize, position: &mut usize, chars: &[char]) -> Result<char, Error> {
    let c = chars
        .get(*position + index)
        .ok_or(Error::UnexpectedEndOfInput)?;

    if c.is_numeric() {
        Ok(*c)
    } else {
        let err = SyntaxError {
            position: *position + index,
            expected: "a number in the range 00..=59".to_string(),
            continues: chars
                .get(*position..*position + 10)
                .or(chars.get(*position..))
                .unwrap()
                .iter()
                .collect::<String>(),
        };
        return Err(Error::Syntax(err));
    }
}

// Matches, parses and returns a collection of weekdays with optional
// modifiers.
fn match_weekdays(
    position: &mut usize,
    chars: &[char],
) -> Result<Vec<(Weekday, Option<WeekdayModifier>)>, Error> {
    let mut tokens = vec![];

    eat_whitespace(position, chars)?;

    let has_braces = match chars.get(*position) {
        Some(c) => {
            if *c == '(' {
                *position += 1;
                true
            } else {
                eat_keyword("on", position, chars)?;
                eat_whitespace(position, chars)?;
                false
            }
        }
        None => {
            return Err(Error::UnexpectedEndOfInput);
        }
    };

    tokens.push(match_weekday(position, chars)?);

    while let Some(ch) = chars.get(*position) {
        if *ch == ',' {
            *position += 1;
            eat_whitespace(position, chars)?;
            tokens.push(match_weekday(position, chars)?);
            continue;
        } else if ch.is_whitespace() {
            if expect_sequence(" and", &position, &chars) {
                eat_whitespace(position, chars)?;
                eat_keyword("and", position, chars)?;
                eat_whitespace(position, chars)?;
                tokens.push(match_weekday(position, chars)?);
                continue;
            } else {
                if has_braces {
                    let err = SyntaxError {
                        position: *position,
                        expected: "'and'".to_string(),
                        continues: chars
                            .get(*position..*position + 10)
                            .or(chars.get(*position..))
                            .unwrap()
                            .iter()
                            .collect::<String>(),
                    };
                    return Err(Error::Syntax(err));
                } else {
                    break;
                }
            }
        } else if *ch == ')' {
            break;
        } else {
            let expected = if has_braces {
                "either ',', ')' or a whitespace".to_string()
            } else {
                "either ',' or a whitespace".to_string()
            };
            let err = SyntaxError {
                position: *position,
                expected,
                continues: chars
                    .get(*position..*position + 10)
                    .or(chars.get(*position..))
                    .unwrap()
                    .iter()
                    .collect::<String>(),
            };
            return Err(Error::Syntax(err));
        }
    }

    if has_braces {
        match chars.get(*position) {
            Some(c) => {
                if *c == ')' {
                    *position += 1;
                } else {
                    let err = SyntaxError {
                        position: *position,
                        expected: "a ')'".to_string(),
                        continues: chars
                            .get(*position..*position + 10)
                            .or(chars.get(*position..))
                            .unwrap()
                            .iter()
                            .collect::<String>(),
                    };
                    return Err(Error::Syntax(err));
                }
            }
            None => {
                return Err(Error::UnexpectedEndOfInput);
            }
        }
    }

    Ok(tokens)
}

// Matches and parses a single weekday with optional modifier.
fn match_weekday(
    position: &mut usize,
    chars: &[char],
) -> Result<(Weekday, Option<WeekdayModifier>), Error> {
    let next = chars
        .get(*position)
        .ok_or(Error::UnexpectedEndOfInput)?
        .clone();

    let mut modifier = None;

    if next.is_numeric() {
        modifier = Some(eat_modifier(position, chars)?);
        eat_whitespace(position, chars)?;
    } else if next.is_alphabetic() && next.is_lowercase() {
        if eat_keyword("the", position, chars).is_ok() {
            eat_whitespace(position, chars)?;
        }
        modifier = Some(eat_modifier(position, chars)?);
        eat_whitespace(position, chars)?;
    }

    let day = if modifier.is_some() {
        eat_weekday(position, chars, true)?
    } else {
        eat_weekday(position, chars, false)?
    };

    return Ok((day, modifier));
}

// Matches and parses the week modifier.
fn match_week(position: &mut usize, chars: &[char]) -> Result<WeekVariant, Error> {
    if *position >= chars.len() {
        return Err(Error::UnexpectedEndOfInput);
    }

    if eat_keyword("in even weeks", position, chars).is_ok() {
        return Ok(WeekVariant::Even);
    } else if eat_keyword("in odd weeks", position, chars).is_ok() {
        return Ok(WeekVariant::Odd);
    } else {
        let err = SyntaxError {
            position: *position,
            expected: "either 'in even weeks' or 'in odd weeks'".to_string(),
            continues: chars
                .get(*position..*position + 10)
                .or(chars.get(*position..))
                .unwrap()
                .iter()
                .collect::<String>(),
        };
        return Err(Error::Syntax(err));
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use time::macros::time;

    #[test]
    fn test_parse_single_block() {
        let spec = ParsedSchedule {
            times: vec![time!(07:30:00), time!(17:00:00), time!(04:00:00)],
            days: Some(vec![
                (Weekday::Monday, None),
                (Weekday::Wednesday, None),
                (Weekday::Friday, Some(WeekdayModifier::Last)),
            ]),
            weeks: Some(WeekVariant::Odd),
        };
        assert_eq!(
            parse("at 07:30 AM, 5 PM and 4 AM on Mondays and Wednesdays and the last Friday in odd weeks"),
            Ok(spec)
        );
    }
}
